tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -360,6 +370,196 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
 	path->rows = compressed_path->rows * DECOMPRESS_CHUNK_BATCH_SIZE;
 }
 
+/*
+ * Calculate the costs for retrieving the decompressed in-order using
+ * a binary heap.
+ */
+static void
+cost_decompress_batch_merge_append(PlannerInfo *root, DecompressChunkPath *dcpath, Path *child_path)
+{
+	Path sort_path; /* dummy for result of cost_sort */
+
+	cost_sort(&sort_path,
+			  root,
+			  dcpath->compressed_pathkeys,
+			  child_path->total_cost,
+			  child_path->rows,
+			  child_path->pathtarget->width,
+			  0.0,
+			  work_mem,
+			  -1);
+
+	/* startup_cost is cost before fetching first tuple */
+	if (sort_path.rows > 0)
+		dcpath->cpath.path.startup_cost = sort_path.total_cost / sort_path.rows;
```

Comment:
I think this is very bad; you may not divide by `numrows` of the `total_cost` of the sort; you need the full sort to be executed to get the first row

or alternatevely you could use the `startup_cost` of the sort...I think that would be the right move here...

Reply:
We have to be compatible with the cost model of the decompress node (see https://github.com/timescale/timescaledb/blob/d3730a4f6ac89612d749d10843def181f7830635/tsl/src/nodes/decompress_chunk/decompress_chunk.c#L356). Therefore, I perform the same calculation here. This is something that could be improved after the cost model of the whole node is improved (maybe in #5505).

Reply:
for DC the basic case is to have a seq scan below; which means:
`total_cost / num_rows ~= (startup + (total-startup)/num_rows)`
so `startup << total_cost`  is more or less true ; but if you have a sort that's not really the case anymore.

for a `Sort`  because `startup ~~ total`  that formula will be off even more....

yeah; we should have #5505 - but it exposed other planning issues

Reply:
note that if this sort would be added via a `Path` you wouldn't even need to complicate with its cost estimation here

Reply:
As discussed, let's change this after we have https://github.com/timescale/timescaledb/pull/5505 (or make it part of this PR and change both cost models at the same time to be compatible).

Reply:
I disagree:
* `startup_cost` of sort is `total_cost` of the child see [here](https://github.com/postgres/postgres/blob/bedc1f0564d12c4a89fc9de3044201166542dc3d/src/backend/optimizer/path/costsize.c#L2113)
* `total_cost >= startup_cost` must be true...
* this formula essentially means that the cost of the 1st row is `1/R * total_cost_of_input`

so in most cases it will be true that `startup_cost(decomp_chunk) << statup_cost(child)` - and it could be off with magnitudes!

in #5505 the (bad) assumption is that the underlying scan is `SeqScan` or `IndexScan`  (which is in most cases true)  - but in this case it will be a `Sort` for sure!

Reply:
I have changed the calculation of the `startup_cost` to take into account that the input needs to be sorted. In general, the cost model could be further optimized. We discussed this in our standup and decided to merge this big PR first and make these kinds of adjustments in smaller follow-up PRs.